### ISSUE TITLE
AnimatedList does not build items correctly when animating their size along with the main axis.

### ISSUE THREAD

**Issue by rasmusir** [flutter/flutter#49293]
## Steps to Reproduce

1. Use this code:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  MyApp({Key key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  final listKey = GlobalKey<AnimatedListState>();
  List<String> items = ["First", "Second", "Thrid", "Fourth"];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          actions: [
            IconButton(
              icon: Icon(Icons.swap_vertical_circle),
              onPressed: () {
                print(" ");
                final toSwap = items.removeLast();
                items.insert(0, toSwap);
                listKey.currentState.insertItem(0);
                listKey.currentState.removeItem(items.length, (context, animation) => itemBuilder(context, toSwap, animation));
              },
            )
          ],
        ),
        body: AnimatedList(
          key: listKey,
          initialItemCount: items.length,
          itemBuilder: (context, index, animation) {
            print("Building $index");
            return itemBuilder(context, items[index], animation);
          },
        ),
      ),
    );
  }

  Widget itemBuilder(BuildContext context, String item, Animation<double> animation) {
    return SizeTransition(
      sizeFactor: animation, 
      axis: Axis.vertical,
      child: Container(
        height: 250,
        child: Text(item),
      ),
    );
  }
}
```
2. Run the application.
3. Scroll to the bottom.
4. Press the Swap Action.

**Expected results:** `Fourth` should be pushed first in the list, using a size transition.

**Actual results:** `Fourth` disappear using a size transition, but it is **not** pushed to the start of the list. It is _gone_. If you reload the application and press `swap` without scrolling to the bottom of the screen it works as expected.

To make this work you might have to change the height of the items, you need to be able to see half the third item on the screen for it to work, it seems to have something to do with when the items are being built, since if they're all visible on the screen it works.

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Linux, locale en_US.UTF-8)
    • Flutter version 1.12.13+hotfix.5 at /home/<username>/development/flutter
    • Framework revision 27321ebbad (6 weeks ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/<username>/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /snap/android-studio/81/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[!] Android Studio (version 3.5)
    • Android Studio at /snap/android-studio/81/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • Pixel • FA77M0302662 • android-arm64 • Android 10 (API 29)

! Doctor found issues in 1 category.
```

</details>

**Comment by rasmusir** [comment]
I also made a Stack Overflow question, in case I'm doing something wrong and this is not the framework. See https://stackoverflow.com/questions/59860889/animatedlist-does-not-seem-to-build-items-correctly-when-animating-their-size-al

**Comment by markusaksli-nc** [comment]
Closing since this seems to be fixed on the latest master `1.22.0-10.0.pre.356`.
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 1.22.0-10.0.pre.356, on Microsoft Windows [Version 10.0.19041.508], locale en-US)
    • Flutter version 1.22.0-10.0.pre.356 at C:\Development\flutter_master
    • Framework revision c412fd9967 (2 hours ago), 2020-09-24 22:47:03 -0700
    • Engine revision 435db234e4
    • Dart version 2.11.0 (build 2.11.0-161.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30320.27
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (6 available)
    • SM G950F (mobile)           • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 arm (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows            • windows-x64    • Microsoft Windows [Version 10.0.19041.508]
    • Web Server (web)            • web-server         • web-javascript • Flutter Tools
    • Chrome (web)                • chrome             • web-javascript • Google Chrome 85.0.4183.121
    • Edge (web)                  • edge               • web-javascript • Microsoft Edge 85.0.564.51

• No issues found!
```
</details>
